### PR TITLE
UI overhaul, consistify, refactoring, mobile/desktop compatibility

### LINKED ISSUE #158: Fix layout of page when view is squashed

### LINKED ISSUE #174: Make UI mobile friendly

### LINKED ISSUE #176: Make UI consistent

### PR DESCRIPTION
**Pull request by JustinPraas**
### Main new features/changes:
1. You can now use a _pagebuilder_ function to build your component into the app. This pagebuilder will handle desktop/mobile layouts for you. The only thing you have to make sure is that your component content itself is mobile _and_ desktop friendly.
2. The pagebuilder function called `buildContent` (See `pagebuilder.tsx`) currently receives 3 parameters:
**2.a.** `headerTitle`: the title that is displayed in the header of the app.
**2.b.** `content`: the actual content of the component that you have to design. If any of your 'fetched' objects are not yet there (a.k.a. they are null) then build the page anyway, but leave this argument at `null`. This will make a spinner appear in the page, which indicates that the page is not ready to load (since for example the `assignmentSets` or `courses` are still null)
**2.c.** `sidebarContent`: the content to be displayed in the sidebar. Leave this `null` if the component does not need a sidebar. 
3. Refactored all pages to incorporate the above described pagebuilder functionality
4. Added JSDOC for the first time! (to pagebuilder file)
5. Refactored the CSS file for 10000x better readability (be wary of merge conflicts! :+1:)
6. Groups manager makes use of sidebar, showing the group members and their comments
7. You can now TRIGGER THE PHONE SIDEBAR from anywhere in your application, which is a must have, for example the comments functionality on the sign off page. This trigger is done by using application state, since otherwise you would have to pass state to sibling components through the pagebuilder, which is only going to make things more complicated. This is a very suitable and clean solution.

Fixes #158
Fixes #174 
Fixes #176

### CODE REVIEW

File: frontend/src/components/pages/login/Login.tsx [hunk 1]
```diff
@@ -68,6 +68,10 @@ class Login extends Component<LoginProps & RouteComponentProps, LoginState> {
         }
     }
 
+    componentWillUnmount() {
+        window.location.reload();
```
Comment (harindup) [inline frontend/src/components/pages/login/Login.tsx]: What's this about?

Comment (JustinPraas) [reply frontend/src/components/pages/login/Login.tsx]: Good thing you noticed. I had some problems where I thought something wasn't loaded after logging in, so I forced a refresh to see if that worked. It didn't work and forgot about it. Will remove this

File: frontend/src/components/comments/CommentThread.tsx [hunk 2]
```diff
@@ -88,9 +88,11 @@ class CommentThread extends Component<CommentThreadProps, CommentThreadState> {
     }
 
     componentDidMount() {
-        const { commentThreadId, needToFetchThread } = this.props;
-        if (needToFetchThread && commentThreadId !== null && this.props.commentThread(commentThreadId) === undefined) {
-            this.props.fetchCommentThread(commentThreadId);
+        const { commentThreadBrief, needToFetchThread } = this.props;
```
Comment (harindup) [inline frontend/src/components/comments/CommentThread.tsx]: The null checks here be `==` checks.

Comment (JustinPraas) [reply frontend/src/components/comments/CommentThread.tsx]: Are changed by the other PR

File: frontend/src/components/comments/CommentThread.tsx [hunk 3]
```diff
@@ -158,19 +160,21 @@ class CommentThread extends Component<CommentThreadProps, CommentThreadState> {
             case (CommentThreadType.Signoff): iconElement = faTasks; break;
         }
 
-        const numberOfComments = this.props.commentThreadId != null ?
-            (this.props.commentThread(this.props.commentThreadId) != null ?
-                this.props.commentThread(this.props.commentThreadId)!.comments.length : 0) : 0;
+        const numberOfComments = this.props.commentThreadBrief !== null ?
```
Comment (harindup) [inline frontend/src/components/comments/CommentThread.tsx]: More null checks needing `==`.

Comment (JustinPraas) [reply frontend/src/components/comments/CommentThread.tsx]: See #200

File: frontend/src/components/comments/CommentThread.tsx (hunk 2 above)
Comment (harindup) [reply frontend/src/components/comments/CommentThread.tsx]: Which one?

File: frontend/src/components/comments/CommentThread.tsx [hunk 4]
```diff
@@ -34,7 +34,7 @@ export enum CommentThreadType {
 interface CommentThreadProps {
     linkedEntityId: number;
     linkedEntityType: CommentThreadType;
-    commentThreadId: number | null;
+    commentThreadBrief: CommentThreadDtoBrief | null;
```
Comment (harindup) [inline frontend/src/components/comments/CommentThread.tsx]: I believe the docs need to be updated for this change?

File: frontend/src/components/navigationBar/NavigationBar.tsx [hunk 5]
```diff
@@ -1,93 +1,118 @@
...
+                                    <img src={logoWhite} />
+                                </Link>
+                            </div> : null}
+                            <div className="mb-auto w-100">
+                                <ListGroup flush className="border-top">
+                                    {
+                                        inCourse ?
+                                            <NavigationBarItem title="Dashboard" icon={faTachometerAlt}
+                                                active={activeTab === ActiveTabEnum.DASHBOARD}
+                                                url={`/courses/${match.params.cid}`} />
```
Comment (harindup) [inline frontend/src/components/navigationBar/NavigationBar.tsx]: These should probably be changed to the && conditional rendering

File: frontend/src/components/navigationBar/NavigationBar.tsx [hunk 6]
```diff
@@ -1,93 +1,118 @@
...
+                            <div className="w-100">
+                                <ListGroup flush className={!onPhone ? "border-top" : ""}>
+                                    <NavigationBarItem title="Courses" icon={faBook}
+                                        active={activeTab === ActiveTabEnum.COURSES} url="/courses" />
+                                    <NavigationBarItem title="Logout" icon={faSignOutAlt} active={false} url="/login" />
+                                </ListGroup>
+                            </div>
+                        </div>
                     </div>
+                    {onPhone ? navbarToggleButton : null}
```
Comment (harindup) [inline frontend/src/components/navigationBar/NavigationBar.tsx]: Use `&&` to render this conditionally.

File: frontend/src/components/pagebuilder.tsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+ * NOTE: Currently renders both mobile and desktop layouts. However, bootstrap handles which
+ * layout needs to be shown using the d-{size}-{display} properties.
+ * @param headerTitle The title that will be displayed in the top of the screen.
+ * @param content The actual content of the body. Needs to be desktop/mobile friendly BEFORE
+ *     it arrives here. Shows a spinner if Null.
+ * @param sidebarContent The content for the sidebar. Needs to be desktop/mobile friendly BEFORE
+ *     it arrives here. If null, does not show the sidebar.
+ * @returns This returns both the desktop and phone components.
+ */
+export const buildContent = (
```
Comment (harindup) [inline frontend/src/components/pagebuilder.tsx]: Give default null values to this function, so that parameters can be omitted.

**Pull request merged**